### PR TITLE
More tightening up of parameter validation during workflow stuff.

### PR DESCRIPTION
**Pull request by jmchilton**
Please review individual commits for details.

Ping @guerler, I believe this is most of what we discussed - what do you think?

### CODE REVIEW

File: lib/galaxy/tools/parameters/basic.py [hunk 1]
```diff
@@ -344,9 +358,8 @@ def from_html( self, value, trans=None, other_values={} ):
...
             return int( value )
         except:
-            if isinstance( value, basestring ):
-                if value.startswith( "$" ) and ( trans is None or trans.workflow_building_mode ):
-                    return value
+            if contains_workflow_parameter(value) and _allow_workflow_parameters_in_context(trans):
+                return value
             elif not value and self.optional:
                 return ""
             if trans is None or trans.workflow_building_mode:
```
Comment (nsoranzo) [inline lib/galaxy/tools/parameters/basic.py]: Also here change `trans is None or trans.workflow_building_mode` to `_allow_workflow_parameters_in_context(trans)`

**Comment by nsoranzo** [comment]
:+1:

**Comment by jmchilton** [comment]
@nsoranzo Done.

**Comment by nsoranzo** [comment]
Thanks @jmchilton, merged! 

I guess we should add a test to check that optional="true" is respected for integer and float tool parameters, I think this is not the first time this has regressed.

**Comment by jmchilton** [comment]
@nsoranzo Agreed, a test case would be great.

**Pull request merged**